hashicorp/dev-portal
# feat: add client-side error view

## 🔗 Relevant links

- [Preview link][preview] 🔎
- [Asana task][task] 🎟️

## 🗒️ What

Adds a dedicated error view for unexpected client-side errors.

## 🤷 Why

To improve the experience and tracking of unexpected client-side errors.

## 🛠️ How

- Adds custom `src/views/error-views/dev-dot-client`, and uses with `ErrorBoundary` in `_app.tsx`
- In `DevDotClient`, loads DataDog script tag
- In `DevDotClient` error view, tracks provided `Error` from `ErrorBoundary` with DataDog RUM
    - We're using the "async CDN" approach, so requires using the global `window.DD_RUM` rather than the library from `npm`. See [Collecting Browser Errors](https://docs.datadoghq.com/real_user_monitoring/browser/collecting_browser_errors/?tab=npm#collect-errors-manually) for details.

## 📸 Design Screenshots

### Client Error View

<img width="1555" alt="client-error-view" src="https://user-images.githubusercontent.com/4624598/174200496-178570f1-bd6d-4ca6-90e1-e9dd8aa9e7fa.png">

### Error in DataDog

<img width="1555" alt="datadog-tracked-error" src="https://user-images.githubusercontent.com/4624598/174201022-442942b4-07ca-459a-a739-0c4dd0796f97.png">

## 🧪 Testing

- [ ] Visit the [temporary preview link, `/test-error-view`][preview]
    - Ensure the client-side error view looks as expected
    - Try triggering an error, and check that it shows up in DataDog

[preview]: https://dev-portal-git-zsclient-error-view-hashicorp.vercel.app/test-error-view
[task]: https://app.asana.com/0/1202114367927919/1202405935832680/f
